packages/neon-core/src/tx/components/Cosigner.ts

```diff
@@ -0,0 +1,118 @@
+import { WitnessScope } from "./WitnessScope";
+import { StringStream, num2hexstring } from "../../u";
+import { deserializeArrayOf, serializeArrayOf } from "../main";
+
+export interface CosignerLike {
+  account: string;
+  scopes: number;
+  allowedContracts: string[];
+  allowedGroups: string[];
+}
+
+export class Cosigner {
+  /**
+   * script hash of cosigner
+   */
+  public account: string;
+  public scopes: WitnessScope;
+  public allowedContracts: string[];
+  public allowedGroups: string[];
+
+  /**
+   * @description This limits maximum number of allowedContracts or allowedGroups here
+   */
+  private readonly MAX_SUB_ITEMS: number = 16;
+
+  public constructor(signer: Partial<CosignerLike>) {
+    const { account, scopes, allowedContracts, allowedGroups } = signer;
+    this.account = account || "";
+    this.scopes = !scopes ? WitnessScope.Global : scopes & 0x11;
+    this.allowedContracts = allowedContracts || [];
+    this.allowedGroups = allowedGroups || [];
+  }
+
+  public get isGlobal() {
+    return this.scopes === WitnessScope.Global;
+  }
+
+  public removeAllScopes() {
+    this.scopes = WitnessScope.Global;
+  }
+
+  public get isCalledByEntry() {
+    return (
+      (this.scopes & WitnessScope.CalledByEntry) === WitnessScope.CalledByEntry
+    );
+  }
+
+  public set isCalledByEntry(yes: boolean) {
+    yes
+      ? (this.scopes |= WitnessScope.CalledByEntry)
+      : (this.scopes &= ~WitnessScope.CalledByEntry);
+  }
+
+  public get isCustomContracts() {
+    return (
+      (this.scopes & WitnessScope.CustomContracts) ===
+      WitnessScope.CustomContracts
+    );
+  }
+
+  public set isCustomContracts(yes: boolean) {
```

Comment:
so this is kinda weird that we have a setter for the scope but not a method to set the allowedContracts. So this leads to cases where the user can:
```
var c = new Cosigner();
c.isCustomContracts= true;
```
which can be an illegal state judging by your export implementation (or a cosigner without scope?)

or the user may forget:
```
var c = new Cosigner();
c.allowedContracts.push(newContract);
```
this is an easy oversight.

Is there any valid use cases of the 2 code examples above? If not, then we can just provide a `addCustomContract` method which will add the contract to the array and amend the scope.